kyo-prelude/shared/src/main/scala/kyo/Batch.scala

```diff
@@ -108,62 +108,63 @@ object Batch:
             ArrowEffect.handle(erasedTag[S], v) {
-                [C] => (input, cont) => Cont(input, v => cont(v.asInstanceOf[C]))
+                [C] =>
+                    (input, cont) =>
+                        val contAny = cont.asInstanceOf[ContAny]
+                        input match
+                            case Call(v, source) => Expanded(v, source.asInstanceOf[SourceAny], contAny)
+                            case v: Seq[C]       => ToExpand(v, contAny)
             }
         end runCont
 
-        case class Expanded(v: Any, source: Source[Any, Any, S], cont: Any => (Cont | A) < (Batch[S] & S & S2))
-        def expand(state: Seq[Cont | A]): Seq[Expanded | A] < (S & S2) =
+        def expand(state: Seq[ToExpand | Expanded | A]): Chunk[Expanded | A] < (S & S2) =
             Kyo.foreach(state) {
-                case Cont(Op.Eval(seq), cont) =>
+                case ToExpand(seq: Seq[Any], cont) =>
                     Kyo.foreach(seq)(v => runCont(cont(v))).map(expand)
-                case Cont(Op.Call(v, source), cont) =>
-                    Seq(Expanded(v, source.asInstanceOf, cont))
-                case a: A @unchecked =>
-                    Seq(a)
-            }.map(_.flatten)
+                case expanded: Expanded @unchecked => Chunk(expanded)
+                case done: A @unchecked            => Chunk(done)
+            }.map(_.flattenChunk)
 
-        def loop(state: Seq[Cont | A]): Seq[A] < (S & S2) =
+        def loop(state: Seq[ToExpand | Expanded | A]): Seq[A] < (S & S2) =
             expand(state).map { expanded =>
-                val (completed, pending) =
-                    expanded.zipWithIndex.partitionMap {
-                        case (e: Expanded @unchecked, idx) => Right((e, idx))
-                        case (a: A @unchecked, idx)        => Left((a, idx))
-                    }
-
-                if pending.isEmpty then
-                    completed.sortBy(_._2).map(_._1)
+                if !expanded.exists((_: @unchecked).isInstanceOf[Expanded]) then
+                    expanded.asInstanceOf[Chunk[A]]
                 else
-                    val grouped = pending.groupBy(_._1.source)
-
-                    Kyo.foreach(grouped.toSeq) { case (source, items) =>
-                        val (expandedItems, indices) = items.unzip
-                        val values                   = expandedItems.map(_.v)
-                        val uniqueValues             = values.distinct
-
-                        source(uniqueValues).map { map =>
-                            Kyo.foreach(expandedItems.zip(indices)) { case (e, idx) =>
-                                runCont(map(e.v).map(e.cont)).map((_, idx))
-                            }
+                    val pending: Map[SourceAny | Unit, Seq[Expanded | A]] =
+                        expanded.groupBy {
+                            case Expanded(_, source, _) => source
+                            case _                      => ()
```

Comment:
I'm creating an artificial source group with unit as the key. This way we can avoid the partition of the collection.